README.markdown
# mysql: bindings to the mysqlclient library

This library is a Haskell binding to the MySQL `mysqlclient` client
library.  It is a fairly faithful, low level library that implements
most of the MySQL client API.  The major departure from the C API is
that in Haskell, resource management is mostly automatic and safe.

This library deliberately avoids the question of providing a "good"
API. Its purpose is to serve as a base upon which higher-level
libraries can be built.

# Licensing

This library is BSD-licensed under the terms of the
[MySQL FOSS License Exception](http://www.mysql.com/about/legal/licensing/foss-exception/).

Since this library links against the GPL-licensed `mysqlclient`
library, a non-open-source application that uses it *may* be subject
to the terms of the GPL.

# To do

* Add support for prepared statements. The prepared statement API is
  huge and of dubious performance value, so it's not currently a
  priority for us. Patches welcome!

# Get involved!

We are happy to receive bug reports, fixes, documentation enhancements,
and other improvements.

Please report bugs via the
[github issue tracker](http://github.com/paul-rouse/mysql/issues).

Master [git repository](http://github.com/paul-rouse/mysql):

* `git clone git://github.com/paul-rouse/mysql.git`

# Authors

This library was written by Bryan O'Sullivan, <bos@serpentine.com>,
to whom all of the credit is due.
It is now being maintained by Paul Rouse, <pyr@doynton.org>.


ChangeLog.md
## 0.2.1

* Work around incorrect test for header-file definition of MYSQL_TYPE_JSON (#45 and 51464fa)

## 0.2.0.1

* Make connection finalizer thread safe (#28).

## 0.2

* Remove obsolete `fieldDefault` from `data Field` (#41).

## 0.1.7.3

* Fix error on certain systems introduced by the change in (#40): some implementations of `mysql_config` do not recognise `--libs-sys`, yet return a zero status for attempts to use it.

## 0.1.7.2

* Update .cabal file

## 0.1.7.1

* Use `mysql_config --libs-sys` in Setup.hs if available (#40)

## 0.1.7

* Add support for JSON type

## 0.1.6

* Fix build for mysql 8

## 0.1.5

* Add Semigroup instance for FieldFlags
* Fix some warnings
* Drop testing under GHC 7.8 / lts-2

## 0.1.4

* Expose mysql_thread_end() as `endThread`

## 0.1.3

* Safer concurrency - see https://ro-che.info/articles/2015-04-17-safe-concurrent-mysql-haskell
* Better support for building against MariaDB (not well tested).
* Additional C binding: mysql_get_server_version().

## 0.1.2.1

* Fix bystestring-valued connectOptions sometimes not being null terminated at the correct place (avoid unsafeUseAsCString).

## 0.1.2

* Fix setup for cabal 1.24
* New maintainer and GitHub location - with many thanks to Bryan O'Sullivan for all of the past work on this module, and for facilitating the transfer of maintenance responsibility.


test/main.hs
{-# LANGUAGE OverloadedStrings #-}

import           Control.Exception                (bracket)
import           Data.Time.Calendar               (fromGregorian)
import           Data.Time.LocalTime              (LocalTime (..),
                                                   TimeOfDay (..))
import           Database.MySQL.Base              (ConnectInfo (..),
                                                   Option (..), bindParams,
                                                   close, connect,
                                                   defaultConnectInfo, execute,
                                                   fetchRow, freeResult,
                                                   initLibrary, initThread,
                                                   prepare, query, storeResult)
import           Database.MySQL.PreparedStatement (Value (..), fetchResults)
import           Test.Hspec

-- This is how to connect to our test database
-- Options with bytestring values are given to partially test #17 and #23
testConn :: ConnectInfo
testConn = defaultConnectInfo {
               connectHost     = "127.0.0.1",
               connectPort = 33306,
               connectUser     = "travis",
               connectPassword     = "esqutest",
               connectDatabase = "esqutest",
               connectOptions  = [
                   InitCommand "SET SESSION sql_mode = 'STRICT_ALL_TABLES';"
                 , ReadDefaultGroup "client"
               ]
           }

-- Only the most cursory test is done at the moment, simply showing that
-- things hang together sufficiently well that we can talk to the database
-- server.
--
main :: IO ()
main = bracket (connect testConn) close $ \conn -> hspec $ do
    describe "Database" $ do
      it "seems to be connected" $ do
        query conn "select 1 + 1"
        result <- storeResult conn
        row <- fetchRow result
        row `shouldBe` [Just "2"]

      it "supports prepared statements" $ do
        stmt <- prepare conn "select 1 where ? = 1"
        bindParams stmt [Long 1]
        execute stmt
        rows <- fetchResults stmt
        rows `shouldBe` [[Just $ LongLong 1]]

      it "supports prepared statements executing multiple times" $ do
        stmt <- prepare conn "select 1 where ? = 1"

        bindParams stmt [Long 2]
        execute stmt
        rows <- fetchResults stmt
        rows `shouldBe` []

        bindParams stmt [Long 1]
        execute stmt
        rows' <- fetchResults stmt
        rows' `shouldBe` [[Just $ LongLong 1]]




Setup.hs

{-# LANGUAGE FlexibleInstances     #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE CPP                   #-}
{- OPTIONS_GHC -Wall #-}

#ifndef MIN_VERSION_Cabal
#define MIN_VERSION_Cabal(x,y,z) 0 
#endif

import Control.Exception (IOException, catch)
import Control.Monad (liftM, msum, sequence)
import Data.List (isPrefixOf, isInfixOf, nub)
import Distribution.PackageDescription
import Distribution.Simple
import Distribution.Simple.LocalBuildInfo
import Distribution.Simple.Program
import Distribution.Verbosity

-- A Cabal 1.16 vs 1.18 compatibility hack, as in 1.18
-- findProgramLocation has a new (unused in this case) parameter.
-- ConstOrId adds this parameter when types say it is mandatory.
class ConstOrId a b where
    constOrId :: a -> b

instance ConstOrId a a where
    constOrId = id

instance ConstOrId a (b -> a) where
    constOrId = const


main = defaultMainWithHooks simpleUserHooks {
  hookedPrograms = [mysqlConfigProgram],

  confHook = \pkg flags -> do
    lbi <- confHook simpleUserHooks pkg flags
    bi  <- mysqlBuildInfo lbi
    return lbi {
      localPkgDescr = updatePackageDescription (Just bi, []) (localPkgDescr lbi)
    }
}

mysqlConfigProgram = (simpleProgram "mysql_config") {
    programFindLocation = \verbosity -> constOrId $ liftM msum $ sequence
#if MIN_VERSION_Cabal(1,24,0)
      [ (findProgramOnSearchPath verbosity [ProgramSearchPathDefault] "mysql_config")
      , (findProgramOnSearchPath verbosity [ProgramSearchPathDefault] "mysql_config5")
      , (findProgramOnSearchPath verbosity [ProgramSearchPathDefault] "mariadb_config")
      ]
#else
      [ (findProgramLocation verbosity "mysql_config")
      , (findProgramLocation verbosity "mysql_config5")
      , (findProgramLocation verbosity "mariadb_config")
      ]
#endif
  }

mysqlBuildInfo :: LocalBuildInfo -> IO BuildInfo
mysqlBuildInfo lbi = do
#if MIN_VERSION_Cabal(2,0,0)
  let mysqlConfig = fmap words . getDbProgramOutput normal
                    mysqlConfigProgram (withPrograms lbi)
#else
  let mysqlConfig = fmap words . rawSystemProgramStdoutConf normal
                    mysqlConfigProgram (withPrograms lbi)
#endif

  include <- mysqlConfig ["--include"]
  libs <- mysqlConfig ["--libs"]
  libsR <- mysqlConfig ["--libs_r"]
  libsSys' <- mysqlConfig ["--libs_sys"] `catch` libsFromError

  -- On some systems, `mysql_config` fails to give an error status even though
  -- it cannot handle `--libs_sys`.  The "Usage:" message lists libraries we do
  -- do not want to include, so recognise it for what it is!
  --
  let libsSys = if null $ filter ("Usage:" `isInfixOf`) libsSys' then libsSys'
                else []

  return emptyBuildInfo {
    extraLibDirs = map (drop 2) . filter ("-L" `isPrefixOf`) $
                   nub (libsSys ++ libs)
  , extraLibs = map (drop 2) . filter ("-l" `isPrefixOf`) .
                filter (/= "-lmygcc") $
                filter (/= "-lmysqlclient_r") $
                nub (libsSys ++ libs ++ libsR)
  , includeDirs = map (drop 2) include
  }
  where
    -- Recover from an error by returning an empty list.
    libsFromError :: IOException -> IO [String]
    libsFromError _ = return []


include/mysql_signals.h
/*
 * Wrappers for MySQL API calls that are known to block and to be
 * vulnerable to interruption by GHC's RTS signals.
 */

#ifndef _mysql_signals_h
#define _mysql_signals_h

#include "mysql.h"

#if !defined(MARIADB_BASE_VERSION) && MYSQL_VERSION_ID >= 80000
typedef char my_bool;
#endif

MYSQL *STDCALL _hs_mysql_real_connect(MYSQL *mysql, const char *host,
				      const char *user,
				      const char *passwd,
				      const char *db,
				      unsigned int port,
				      const char *unix_socket,
				      unsigned long clientflag);
void STDCALL _hs_mysql_close(MYSQL *sock);
int STDCALL _hs_mysql_ping(MYSQL *mysql);
int STDCALL _hs_mysql_real_query(MYSQL *mysql, const char *q,
				 unsigned long length);
const char *STDCALL _hs_mysql_stat(MYSQL *mysql);
my_bool STDCALL _hs_mysql_commit(MYSQL * mysql);
my_bool STDCALL _hs_mysql_rollback(MYSQL * mysql);
my_bool STDCALL _hs_mysql_autocommit(MYSQL * mysql, my_bool auto_mode);
my_bool STDCALL _hs_mysql_change_user(MYSQL *mysql, const char *user,
				      const char *passwd, const char *db);
int STDCALL _hs_mysql_select_db(MYSQL *mysql, const char *db);
MYSQL_FIELD *STDCALL _hs_mysql_fetch_field(MYSQL_RES *result);
MYSQL_ROW STDCALL _hs_mysql_fetch_row(MYSQL_RES *result);
unsigned long *STDCALL _hs_mysql_fetch_lengths(MYSQL_RES *result);
MYSQL_RES *STDCALL _hs_mysql_store_result(MYSQL *mysql);
MYSQL_RES *STDCALL _hs_mysql_use_result(MYSQL *mysql);
void STDCALL _hs_mysql_free_result(MYSQL_RES *result);
int STDCALL _hs_mysql_next_result(MYSQL *mysql);
void STDCALL _hs_mysql_close(MYSQL *sock);

#endif /* _mysql_signals_h */


Database/MySQL/Base.hs
{-# LANGUAGE DeriveDataTypeable, ForeignFunctionInterface, RecordWildCards #-}

-- |
-- Module:      Database.MySQL.Base
-- Copyright:   (c) 2011 MailRank, Inc.
-- License:     BSD3
-- Maintainer:  Paul Rouse <pyr@doynton.org>
-- Stability:   experimental
-- Portability: portable
--
-- A low-level client library for the MySQL database, implemented as
-- bindings to the C @mysqlclient@ API.
--
-- The C library is thread-safe, but uses thread-local state.  Therefore,
-- if these bindings are used in a multi-threaded program, "bound" threads
-- should be used (see "Control.Concurrent").  In addition, explicit calls
-- to 'initLibrary', and possibly 'initThread' and 'endThread' may be needed
-- in a multi-threaded program.

module Database.MySQL.Base
    (
    -- * Licensing
    -- $license
    -- * Resource management
    -- $mgmt
    -- * Types
      ConnectInfo(..)
    , SSLInfo(..)
    , Seconds
    , Protocol(..)
    , Option(..)
    , defaultConnectInfo
    , defaultSSLInfo
    , Connection
    , Result
    , Type(..)
    , Row
    , MySQLError(errFunction, errNumber, errMessage)
    -- * Connection management
    , connect
    , close
    , autocommit
    , ping
    , changeUser
    , selectDB
    , setCharacterSet
    -- ** Connection information
    , threadId
    , serverInfo
    , hostInfo
    , protocolInfo
    , characterSet
    , sslCipher
    , serverStatus
    -- * Querying
    , query
    , insertID
    -- ** Escaping
    , escape
    -- ** Results
    , fieldCount
    , affectedRows
    -- * Working with results
    , isResultValid
    , freeResult
    , storeResult
    , useResult
    , fetchRow
    , fetchFields
    , dataSeek
    , rowSeek
    , rowTell
    -- ** Multiple results
    , nextResult
    -- * Transactions
    , commit
    , rollback
    -- * General information
    , clientInfo
    , clientVersion
    -- * Concurrency
    , initLibrary
    , initThread
    , endThread
    -- * Prepared Statements
    , prepare
    , Statement.bindParams
    , Statement.execute
    ) where

import Control.Applicative ((<$>), (<*>))
import Control.Concurrent (rtsSupportsBoundThreads, runInBoundThread)
import Control.Exception (Exception, throw)
import Control.Monad (forM_, unless, when)
import Data.ByteString.Char8 ()
import Data.ByteString.Internal (ByteString, create, createAndTrim, memcpy)
import Data.ByteString.Unsafe (unsafeUseAsCStringLen)
import Data.IORef (IORef, atomicModifyIORef, newIORef, readIORef, writeIORef)
import Data.Int (Int64)
import Data.List (foldl')
import Data.Typeable (Typeable)
import Data.Word (Word, Word16, Word64)
import Data.Text (Text)
import qualified Database.MySQL.PreparedStatement as Statement
import Database.MySQL.Base.C
import Database.MySQL.Base.Types
import Foreign.C.String (CString, peekCString, withCString)
import Foreign.C.Types (CULong)
import Foreign.Concurrent (newForeignPtr)
import Foreign.ForeignPtr hiding (newForeignPtr)
import Foreign.Marshal.Array (peekArray)
import Foreign.Ptr (Ptr, castPtr, nullPtr)
import System.IO.Unsafe (unsafePerformIO)
import System.Mem.Weak (Weak, deRefWeak, mkWeakPtr)

-- $license
--
-- /Important licensing note/: This library is BSD-licensed under the
-- terms of the MySQL FOSS License Exception
-- <http://www.mysql.com/about/legal/licensing/foss-exception/>.
--
-- Since this library links against the GPL-licensed @mysqlclient@
-- library, a non-open-source application that uses it /may/ be
-- subject to the terms of the GPL.

-- $mgmt
--
-- Our rules for managing 'Connection' and 'Result' values are
-- unfortunately complicated, thanks to MySQL's lifetime rules.
--
-- At the C @libmysqlclient@ level, a single @MYSQL@ connection may
-- cause multiple @MYSQL_RES@ result values to be created over the
-- course of multiple queries, but only one of these @MYSQL_RES@
-- values may be alive at a time.  The programmer is responsible for
-- knowing when to call @mysql_free_result@.
--
-- Meanwhile, up in Haskell-land, we'd like both 'Connection' and
-- 'Result' values to be managed either manually or automatically. In
-- particular, we want finalizers to tidy up after a messy programmer,
-- and we'd prefer it if people didn't need to be mindful of calling
-- @mysql_free_result@. This means that we must wrestle with the
-- lifetime rules. An obvious approach would be to use some monad and
-- type magic to enforce those rules, but then we'd end up with an
-- awkward API.
--
-- Instead, we allow 'Result' values to stay alive for arbitrarily
-- long times, while preserving the right to mark them as
-- invalid. When a @Result@ is marked invalid, its associated
-- @MYSQL_RES@ is freed, and can no longer be used.
--
-- Since all functions over @Result@ values are in the 'IO' monad, we
-- don't risk disrupting pure code by introducing this notion of
-- invalidity. If code tries to use an invalid @Result@, a
-- 'MySQLError' will be thrown. This should /not/ occur in normal
-- code, so there should be no need to use 'isResultValid' to test a
-- @Result@ for validity.
--
-- Each of the following functions will invalidate a 'Result':
--
-- * 'close'
--
-- * 'freeResult'
--
-- * 'nextResult'
--
-- * 'storeResult'
--
-- * 'useResult'
--
-- A 'Result' must be able to keep a 'Connection' alive so that a
-- streaming @Result@ constructed by 'useResult' can continue to pull
-- data from the server, but a @Connection@ must (a) be able to cause
-- the @MYSQL_RES@ behind a @Result@ to be deleted at a moment's notice,
-- while (b) not artificially prolonging the life of either the @Result@
-- or its @MYSQL_RES@.

data ConnectInfo = ConnectInfo {
      connectHost :: String
    , connectPort :: Word16
    , connectUser :: String
    , connectPassword :: String
    , connectDatabase :: String
    , connectOptions :: [Option]
    , connectPath :: FilePath
    , connectSSL :: Maybe SSLInfo
    } deriving (Eq, Read, Show, Typeable)

data SSLInfo = SSLInfo {
      sslKey :: FilePath
    , sslCert :: FilePath
    , sslCA :: FilePath
    , sslCAPath :: FilePath
    , sslCiphers :: String -- ^ Comma-separated list of cipher names.
    } deriving (Eq, Read, Show, Typeable)

-- | The constructors of @MySQLError@ are not currently exported, but they
--   have a consistent set of field names which are exported.  These fields are:
--
--   >  errFunction :: String
--   >  errNumber   :: Int
--   >  errMessage  :: String
--
data MySQLError = ConnectionError {
      errFunction :: String
    , errNumber :: Int
    , errMessage :: String
    } | ResultError {
      errFunction :: String
    , errNumber :: Int
    , errMessage :: String
    } deriving (Eq, Show, Typeable)

instance Exception MySQLError

-- | Connection to a MySQL database.
data Connection = Connection {
      connFP :: ForeignPtr MYSQL
    , connClose :: IO ()
    , connResult :: IORef (Maybe (Weak Result))
    } deriving (Typeable)

-- | Result of a database query.
data Result = Result {
      resFP :: ForeignPtr MYSQL_RES
    , resFields :: {-# UNPACK #-} !Int
    , resConnection :: Connection
    , resValid :: IORef Bool
    , resFetchFields :: Ptr MYSQL_RES -> IO (Ptr Field)
    , resFetchRow :: Ptr MYSQL_RES -> IO MYSQL_ROW
    , resFetchLengths :: Ptr MYSQL_RES -> IO (Ptr CULong)
    , resFreeResult :: Ptr MYSQL_RES -> IO ()
    } | EmptyResult
  deriving (Typeable)

-- | A row cursor, used by 'rowSeek' and 'rowTell'.
newtype Row = Row MYSQL_ROW_OFFSET
  deriving (Typeable)

-- | Default information for setting up a connection.
--
-- Defaults are as follows:
--
-- * Server on @localhost@
--
-- * User @root@
--
-- * No password
--
-- * Database @test@
--
-- * Character set @utf8@
--
-- Use as in the following example:
--
-- > connect defaultConnectInfo { connectHost = "db.example.com" }
defaultConnectInfo :: ConnectInfo
defaultConnectInfo = ConnectInfo {
                       connectHost = "localhost"
                     , connectPort = 3306
                     , connectUser = "root"
                     , connectPassword = ""
                     , connectDatabase = "test"
                     , connectOptions = [CharsetName "utf8"]
                     , connectPath = ""
                     , connectSSL = Nothing
                     }

-- | Default (empty) information for setting up an SSL connection.
defaultSSLInfo :: SSLInfo
defaultSSLInfo = SSLInfo {
                   sslKey = ""
                 , sslCert = ""
                 , sslCA = ""
                 , sslCAPath = ""
                 , sslCiphers = ""
                 }

-- | Connect to a database.
connect :: ConnectInfo -> IO Connection
connect ConnectInfo{..} = do
  closed <- newIORef False
  ptr0 <- mysql_init nullPtr
  case connectSSL of
    Nothing -> return ()
    Just SSLInfo{..} -> withString sslKey $ \ckey ->
                         withString sslCert $ \ccert ->
                          withString sslCA $ \cca ->
                           withString sslCAPath $ \ccapath ->
                            withString sslCiphers $ \ccipher ->
                             mysql_ssl_set ptr0 ckey ccert cca ccapath ccipher
                             >> return ()
  forM_ connectOptions $ \opt -> do
    r <- mysql_options ptr0 opt
    unless (r == 0) $ connectionError_ "connect" ptr0
  let flags = foldl' (+) 0 . map toConnectFlag $ connectOptions
  ptr <- withString connectHost $ \chost ->
          withString connectUser $ \cuser ->
           withString connectPassword $ \cpass ->
            withString connectDatabase $ \cdb ->
             withString connectPath $ \cpath ->
               mysql_real_connect ptr0 chost cuser cpass cdb
                                  (fromIntegral connectPort) cpath flags
  when (ptr == nullPtr) $
    connectionError_ "connect" ptr0
  res <- newIORef Nothing
  let realClose = do
        cleanupConnResult res
        wasClosed <- atomicModifyIORef closed $ \prev -> (True, prev)
        unless wasClosed $ mysql_close ptr
  -- In general, the user of this library is responsible for dealing with thread
  -- safety. However, the programmer has no control over the OS thread
  -- finalizers are run from so we use 'runInBoundThread' and 'initThread' here.
  let myRunInBoundThread = if rtsSupportsBoundThreads then runInBoundThread else id
  fp <- newForeignPtr ptr (myRunInBoundThread $ initThread >> realClose)
  return Connection {
               connFP = fp
             , connClose = realClose
             , connResult = res
             }

-- | Delete the 'MYSQL_RES' behind a 'Result' immediately, and mark
-- the 'Result' as invalid.
cleanupConnResult :: IORef (Maybe (Weak Result)) -> IO ()
cleanupConnResult res = do
  prev <- readIORef res
  case prev of
    Nothing -> return ()
    Just w -> maybe (return ()) freeResult =<< deRefWeak w

-- | Close a connection, and mark any outstanding 'Result' as
-- invalid.
close :: Connection -> IO ()
close = connClose
{-# INLINE close #-}

ping :: Connection -> IO ()
ping conn = withConn conn $ \ptr -> mysql_ping ptr >>= check "ping" conn

threadId :: Connection -> IO Word
threadId conn = fromIntegral <$> withConn conn mysql_thread_id

serverInfo :: Connection -> IO String
serverInfo conn = withConn conn $ \ptr ->
                  peekCString =<< mysql_get_server_info ptr

hostInfo :: Connection -> IO String
hostInfo conn = withConn conn $ \ptr ->
                peekCString =<< mysql_get_host_info ptr

protocolInfo :: Connection -> IO Word
protocolInfo conn = withConn conn $ \ptr ->
                    fromIntegral <$> mysql_get_proto_info ptr

setCharacterSet :: Connection -> String -> IO ()
setCharacterSet conn cs =
  withCString cs $ \ccs ->
    withConn conn $ \ptr ->
        mysql_set_character_set ptr ccs >>= check "setCharacterSet" conn

characterSet :: Connection -> IO String
characterSet conn = withConn conn $ \ptr ->
  peekCString =<< mysql_character_set_name ptr

sslCipher :: Connection -> IO (Maybe String)
sslCipher conn = withConn conn $ \ptr ->
  withPtr peekCString =<< mysql_get_ssl_cipher ptr

serverStatus :: Connection -> IO String
serverStatus conn = withConn conn $ \ptr -> do
  st <- mysql_stat ptr
  checkNull "serverStatus" conn st
  peekCString st

clientInfo :: String
clientInfo = unsafePerformIO $ peekCString mysql_get_client_info
{-# NOINLINE clientInfo #-}

clientVersion :: Word
clientVersion = fromIntegral mysql_get_client_version
{-# NOINLINE clientVersion #-}

-- | Turn autocommit on or off.
--
-- By default, MySQL runs with autocommit mode enabled. In this mode,
-- as soon as you modify a table, MySQL stores your modification
-- permanently.
autocommit :: Connection -> Bool -> IO ()
autocommit conn onOff = withConn conn $ \ptr ->
   mysql_autocommit ptr b >>= check "autocommit" conn
 where b = if onOff then 1 else 0

changeUser :: Connection -> String -> String -> Maybe String -> IO ()
changeUser conn user pass mdb =
  withCString user $ \cuser ->
   withCString pass $ \cpass ->
    withMaybeString mdb $ \cdb ->
     withConn conn $ \ptr ->
      mysql_change_user ptr cuser cpass cdb >>= check "changeUser" conn

selectDB :: Connection -> String -> IO ()
selectDB conn db =
  withCString db $ \cdb ->
    withConn conn $ \ptr ->
      mysql_select_db ptr cdb >>= check "selectDB" conn

query :: Connection -> ByteString -> IO ()
query conn q = withConn conn $ \ptr ->
  unsafeUseAsCStringLen q $ \(p,l) ->
  mysql_real_query ptr p (fromIntegral l) >>= check "query" conn

-- | Return the value generated for an @AUTO_INCREMENT@ column by the
-- previous @INSERT@ or @UPDATE@ statement.
--
-- See <http://dev.mysql.com/doc/refman/5.5/en/mysql-insert-id.html>
insertID :: Connection -> IO Word64
insertID conn = fromIntegral <$> (withConn conn $ mysql_insert_id)

-- | Return the number of fields (columns) in a result.
--
-- * If 'Left' 'Connection', returns the number of columns for the most
--   recent query on the connection.
--
-- * For 'Right' 'Result', returns the number of columns in each row
--   of this result.
--
-- The number of columns may legitimately be zero.
fieldCount :: Either Connection Result -> IO Int
fieldCount (Right EmptyResult) = return 0
fieldCount (Right res)         = return (resFields res)
fieldCount (Left conn)         =
    withConn conn $ fmap fromIntegral . mysql_field_count

affectedRows :: Connection -> IO Int64
affectedRows conn = withConn conn $ fmap fromIntegral . mysql_affected_rows

-- | Retrieve a complete result.
--
-- Any previous outstanding 'Result' is first marked as invalid.
storeResult :: Connection -> IO Result
storeResult = frobResult "storeResult" mysql_store_result
              mysql_fetch_fields_nonblock
              mysql_fetch_row_nonblock
              mysql_fetch_lengths_nonblock
              mysql_free_result_nonblock

-- | Initiate a row-by-row retrieval of a result.
--
-- Any previous outstanding 'Result' is first marked as invalid.
useResult :: Connection -> IO Result
useResult = frobResult "useResult" mysql_use_result
            mysql_fetch_fields
            mysql_fetch_row
            mysql_fetch_lengths
            mysql_free_result

frobResult :: String
           -> (Ptr MYSQL -> IO (Ptr MYSQL_RES))
           -> (Ptr MYSQL_RES -> IO (Ptr Field))
           -> (Ptr MYSQL_RES -> IO MYSQL_ROW)
           -> (Ptr MYSQL_RES -> IO (Ptr CULong))
           -> (Ptr MYSQL_RES -> IO ())
           -> Connection -> IO Result
frobResult func frob fetchFieldsFunc fetchRowFunc fetchLengthsFunc
           myFreeResult conn =
  withConn conn $ \ptr -> do
    cleanupConnResult (connResult conn)
    res <- frob ptr
    fields <- mysql_field_count ptr
    valid <- newIORef True
    if res == nullPtr
      then if fields == 0
           then return EmptyResult
           else connectionError func conn
      else do
        fp <- newForeignPtr res $ freeResult_ valid myFreeResult res
        let ret = Result {
                    resFP = fp
                  , resFields = fromIntegral fields
                  , resConnection = conn
                  , resValid = valid
                  , resFetchFields = fetchFieldsFunc
                  , resFetchRow = fetchRowFunc
                  , resFetchLengths = fetchLengthsFunc
                  , resFreeResult = myFreeResult
                  }
        weak <- mkWeakPtr ret (Just (freeResult_ valid myFreeResult res))
        writeIORef (connResult conn) (Just weak)
        return ret

-- | Immediately free the @MYSQL_RES@ value associated with this
-- 'Result', and mark the @Result@ as invalid.
freeResult :: Result -> IO ()
freeResult Result{..}  = withForeignPtr resFP $
                         freeResult_ resValid resFreeResult
freeResult EmptyResult = return ()

-- | Check whether a 'Result' is still valid, i.e. backed by a live
-- @MYSQL_RES@ value.
isResultValid :: Result -> IO Bool
isResultValid Result{..}  = readIORef resValid
isResultValid EmptyResult = return False

freeResult_ :: IORef Bool -> (Ptr MYSQL_RES -> IO ()) -> Ptr MYSQL_RES -> IO ()
freeResult_ valid free ptr = do
  wasValid <- atomicModifyIORef valid $ \prev -> (False, prev)
  when wasValid $ free ptr

fetchRow :: Result -> IO [Maybe ByteString]
fetchRow res@Result{..}  = withRes "fetchRow" res $ \ptr -> do
  rowPtr <- resFetchRow ptr
  if rowPtr == nullPtr
    then return []
    else do
      lenPtr <- resFetchLengths ptr
      checkNull "fetchRow" resConnection lenPtr
      let go len = withPtr $ \colPtr ->
                   create (fromIntegral len) $ \d ->
                   memcpy d (castPtr colPtr) (fromIntegral len)
      sequence =<< zipWith go <$> peekArray resFields lenPtr
                              <*> peekArray resFields rowPtr
fetchRow EmptyResult = return []

fetchFields :: Result -> IO [Field]
fetchFields res@Result{..} = withRes "fetchFields" res $ \ptr -> do
  peekArray resFields =<< resFetchFields ptr
fetchFields EmptyResult    = return []

dataSeek :: Result -> Int64 -> IO ()
dataSeek res row = withRes "dataSeek" res $ \ptr ->
  mysql_data_seek ptr (fromIntegral row)

rowTell :: Result -> IO Row
rowTell res = withRes "rowTell" res $ \ptr ->
  Row <$> mysql_row_tell ptr

rowSeek :: Result -> Row -> IO Row
rowSeek res (Row row) = withRes "rowSeek" res $ \ptr ->
  Row <$> mysql_row_seek ptr row

-- | Read the next statement result. Returns 'True' if another result
-- is available, 'False' otherwise.
--
-- This function marks the current 'Result' as invalid, if one exists.
nextResult :: Connection -> IO Bool
nextResult conn = withConn conn $ \ptr -> do
  cleanupConnResult (connResult conn)
  i <- mysql_next_result ptr
  case i of
    0  -> return True
    -1 -> return False
    _  -> connectionError "nextResult" conn

-- | Commit the current transaction.
commit :: Connection -> IO ()
commit conn = withConn conn $ \ptr ->
              mysql_commit ptr >>= check "commit" conn

-- | Roll back the current transaction.
rollback :: Connection -> IO ()
rollback conn = withConn conn $ \ptr ->
                mysql_rollback ptr >>= check "rollback" conn

escape :: Connection -> ByteString -> IO ByteString
escape conn bs = withConn conn $ \ptr ->
  unsafeUseAsCStringLen bs $ \(p,l) ->
    createAndTrim (l*2 + 1) $ \to ->
      fromIntegral <$> mysql_real_escape_string ptr (castPtr to) p
                                                (fromIntegral l)

withConn :: Connection -> (Ptr MYSQL -> IO a) -> IO a
withConn conn = withForeignPtr (connFP conn)

-- | Call @mysql_library_init@
--
-- A single-threaded program can rely on an implicit initialisation done
-- when making the first connection, but a multi-threaded one should call
-- 'initLibrary' separately, and it should be done before other threads
-- might call into this library, since this function is not thread-safe.
-- See <https://ro-che.info/articles/2015-04-17-safe-concurrent-mysql-haskell>
-- and <https://dev.mysql.com/doc/refman/5.7/en/c-api-threaded-clients.html>
-- for details.
initLibrary :: IO ()
initLibrary = do
  r <- mysql_library_init 0 nullPtr nullPtr
  if r == 0
    then return ()
    else throw $ ConnectionError "initLibrary" (-1)
      "mysql_library_init failed"

-- | Call @mysql_thread_init@
--
-- Again a single-threaded program does not need to call this explicitly.  Even
-- in a multi-threaded one, if each connection is made, used, and destroyed
-- in a single thread, it is sufficient to rely on the 'connect' call to do
-- an implicit thread initialisation.  But in other cases, for example when
-- using a connection pool, each thread requires explicit initialisation.
-- See <https://ro-che.info/articles/2015-04-17-safe-concurrent-mysql-haskell>
-- and <https://dev.mysql.com/doc/refman/5.7/en/c-api-threaded-clients.html>
-- for details.
initThread :: IO ()
initThread = do
  r <- mysql_thread_init
  if r == 0
    then return ()
    else throw $ ConnectionError "initThread" (-1)
      "mysql_thread_init failed"

-- | Call @mysql_thread_end@
--
-- This is needed at thread exit to avoid a memory leak, except when using
-- a non-debug build of at least version 5.7.9 of the MySQL library.
-- See <https://dev.mysql.com/doc/refman/5.7/en/mysql-thread-end.html>.
-- The threads in question are the /OS threads/, so calling this function
-- is likely to be important when using large numbers of bound threads (see
-- "Control.Concurrent").  Unbound threads - those created with 'forkIO' and
-- friends - share a small number of OS threads, so in those it is hard to
-- call this function safely, and there is little benefit in doing so, but in
-- any case using this library in unbound threads is not recommended  (see
-- <https://ro-che.info/articles/2015-04-17-safe-concurrent-mysql-haskell>).
endThread :: IO ()
endThread = mysql_thread_end

withRes :: String -> Result -> (Ptr MYSQL_RES -> IO a) -> IO a
withRes func res act = do
  valid <- readIORef (resValid res)
  unless valid . throw $ ResultError func 0 "result is no longer usable"
  withForeignPtr (resFP res) act

withString :: String -> (CString -> IO a) -> IO a
withString [] act = act nullPtr
withString xs act = withCString xs act

withMaybeString :: Maybe String -> (CString -> IO a) -> IO a
withMaybeString Nothing act = act nullPtr
withMaybeString (Just xs) act = withCString xs act

check :: (Eq a, Num a) => String -> Connection -> a -> IO ()
check func conn r = unless (r == 0) $ connectionError func conn
{-# INLINE check #-}

checkNull :: String -> Connection -> Ptr a -> IO ()
checkNull func conn p = when (p == nullPtr) $ connectionError func conn
{-# INLINE checkNull #-}

withPtr :: (Ptr a -> IO b) -> Ptr a -> IO (Maybe b)
withPtr act p | p == nullPtr = return Nothing
              | otherwise    = Just <$> act p

connectionError :: String -> Connection -> IO a
connectionError func conn = withConn conn $ connectionError_ func

connectionError_ :: String -> Ptr MYSQL -> IO a
connectionError_ func ptr =do
  errno <- mysql_errno ptr
  msg <- peekCString =<< mysql_error ptr
  throw $ ConnectionError func (fromIntegral errno) msg

prepare :: Connection -> Text -> IO Statement.Statement
prepare conn q = do
    Statement.prepare (connFP conn) q
